docker-app/qfieldcloud/core/middleware/qgis_auth.py

```diff
@@ -0,0 +1,129 @@
+
+    On subsequent requests QGIS sends back those session cookies, and
+    authentication does not need to be performed again.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        """Detect and extract HTTP headers intended for SSO auth.
+
+        Because this code gets called for *every* single request, we need to
+        be quite defensive here. For reasons of performance and robustness.
+
+        We need to abort as early as possible if we can determine that the
+        request is not intended for SSO auth. And we must not cause any
+        unhandled exceptions that could interfere with normal requests.
+        """
+        if request.user.is_authenticated:
+            # User already has a session
+            return self.get_response(request)
+
+        id_token = self.get_id_token(request)
+        access_token = self.get_access_token(request)
+
+        if not any([id_token, access_token]):
+            # No tokens found, nothing for us to do
+            return self.get_response(request)
+
+        # Require QFC clients to send a header to explicitly specify the IDP
+        idp_id = request.headers.get("X-QFC-IDP-ID")
+        if not idp_id:
+            return self.get_response(request)
+
+        provider = self.get_provider(request, idp_id)
+        if not provider:
+            # Could not determine the provider, bail
+            return self.get_response(request)
+
+        # Delegate the actual authentication to django-allauth.
+        #
+        # We do this by emulating the parameters for django-allauth's
+        # OAuth2Adapter.complete_login() method in the way that it expects them.
+        #
+        # Specifically, a SocialToken instance with the access token, and a
+        # `response` dictionary with the ID token (if available).
+        token = SocialToken(token=access_token)
+        token_response = {}
+        if id_token:
+            token_response["id_token"] = id_token
+
+        # This verifies the token, determines attributes like 'uid' and 'email',
+        # and prepares a SocialLogin instance.
+        oauth_adapter = provider.get_oauth2_adapter(request)
+        social_login = oauth_adapter.complete_login(
+            request, provider.app, token, response=token_response
+        )
+
+        # This performs signup of the a user, if necessary, authenticates
+        # the user, links the social account, and creates a session.
+        complete_social_login(request, social_login)
+
+        logger.info("Authenticated user: %s" % request.user)
+        return self.get_response(request)
+
+    def get_id_token(self, request):
+        return request.headers.get("X-QFC-ID-Token")
+
+    def get_access_token(self, request):
+        auth_header = request.headers.get("Authorization")
+
+        if not (auth_header and auth_header.startswith("Bearer ")):
+            return
```

Comment:
```suggestion
            return None
```

Or maybe it's implicit ?